File: habit_reminder.go
// habit_reminder.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"
	"github.com/google/uuid"
)

type Habit struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	FrequencyDays int   `json:"frequency_days"`
	LastDone     string `json:"last_done"`
}

func (h *Habit) NextDueDate() time.Time {
	if h.LastDone == "" {
		return time.Now()
	}
	last, _ := time.Parse(time.RFC3339, h.LastDone)
	return last.AddDate(0, 0, h.FrequencyDays)
}

func (h *Habit) IsDue() bool {
	due := h.NextDueDate()
	return due.Before(time.Now()) || due.Equal(time.Now())
}

func (h *Habit) MarkDone() {
	h.LastDone = time.Now().Format(time.RFC3339)
}

type Tracker struct {
	Habits []Habit `json:"habits"`
}

var dataFile = "habits.json"

func (t *Tracker) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, t)
}

func (t *Tracker) save() {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (t *Tracker) add(name string, freq int) {
	habit := Habit{
		ID:           uuid.New().String()[:8],
		Name:         name,
		FrequencyDays: freq,
	}
	t.Habits = append(t.Habits, habit)
	t.save()
	fmt.Printf("✅ Habit added: %s (ID: %s, frequency: %d days)\n", name, habit.ID, freq)
}

func (t *Tracker) list() {
	if len(t.Habits) == 0 {
		fmt.Println("No habits.")
		return
	}
	fmt.Println("\n📋 Your Habits:")
	for _, h := range t.Habits {
		due := h.NextDueDate()
		daysUntil := int(due.Sub(time.Now()).Hours() / 24)
		status := ""
		if daysUntil < 0 {
			status = "🔔 DUE!"
		} else if daysUntil == 0 {
			status = "⏰ due today"
		} else if daysUntil == 1 {
			status = "due tomorrow"
		} else {
			status = fmt.Sprintf("due in %d days", daysUntil)
		}
		last := h.LastDone
		if last == "" {
			last = "never"
		} else {
			last = last[:10]
		}
		fmt.Printf("  %s: %s (every %d day%s) – last: %s, next: %s (%s)\n",
			h.ID, h.Name, h.FrequencyDays, func() string {
				if h.FrequencyDays > 1 { return "s" } else { return "" }
			}(), last, due.Format("2006-01-02"), status)
	}
}

func (t *Tracker) check() {
	due := []Habit{}
	for _, h := range t.Habits {
		if h.IsDue() {
			due = append(due, h)
		}
	}
	if len(due) == 0 {
		fmt.Println("✅ No habits due today. Well done!")
		return
	}
	fmt.Println("\n🔔 Due habits:")
	reader := bufio.NewReader(os.Stdin)
	for _, h := range due {
		fmt.Printf("  %s: %s (frequency: %d days)\n", h.ID, h.Name, h.FrequencyDays)
		fmt.Printf("  Mark '%s' as done? (y/n): ", h.Name)
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(strings.ToLower(input))
		if input == "y" {
			for i := range t.Habits {
				if t.Habits[i].ID == h.ID {
					t.Habits[i].MarkDone()
					break
				}
			}
			fmt.Printf("  ✅ Habit '%s' marked as done.\n", h.Name)
		} else {
			fmt.Printf("  ⏳ Habit '%s' skipped.\n", h.Name)
		}
	}
	t.save()
}

func (t *Tracker) remove(id string) {
	for i, h := range t.Habits {
		if h.ID == id {
			name := h.Name
			t.Habits = append(t.Habits[:i], t.Habits[i+1:]...)
			t.save()
			fmt.Printf("✅ Habit '%s' removed.\n", name)
			return
		}
	}
	fmt.Printf("❌ Habit with ID %s not found.\n", id)
}

func (t *Tracker) watch(interval int) {
	fmt.Printf("🔔 Habit Reminder – checking every %d minute(s). Press Ctrl+C to stop.\n", interval)
	ticker := time.NewTicker(time.Duration(interval) * time.Minute)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			due := []Habit{}
			for _, h := range t.Habits {
				if h.IsDue() {
					due = append(due, h)
				}
			}
			if len(due) > 0 {
				fmt.Printf("\n🔔 Due habits at %s:\n", time.Now().Format("15:04"))
				for _, h := range due {
					fmt.Printf("  %s: %s\n", h.ID, h.Name)
				}
			} else {
				fmt.Print(".")
			}
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: habit_reminder <command> [options]")
		return
	}
	tracker := &Tracker{}
	tracker.load()
	cmd := os.Args[1]
	switch cmd {
	case "add":
		addCmd := flag.NewFlagSet("add", flag.ExitOnError)
		name := addCmd.String("name", "", "")
		freq := addCmd.Int("frequency", 1, "")
		addCmd.Parse(os.Args[2:])
		if *name == "" && len(addCmd.Args()) > 0 {
			*name = addCmd.Args()[0]
		}
		if *name == "" {
			fmt.Println("add requires a name")
			return
		}
		tracker.add(*name, *freq)
	case "list":
		tracker.list()
	case "check":
		tracker.check()
	case "remove":
		if len(os.Args) < 3 {
			fmt.Println("remove <id>")
			return
		}
		tracker.remove(os.Args[2])
	case "watch":
		watchCmd := flag.NewFlagSet("watch", flag.ExitOnError)
		interval := watchCmd.Int("interval", 5, "")
		watchCmd.Parse(os.Args[2:])
		tracker.watch(*interval)
	default:
		fmt.Println("Unknown command. Use add, list, check, remove, watch.")
	}
}

File: habit_reminder.cpp
// habit_reminder.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <ctime>
#include <iomanip>
#include <random>
#include <nlohmann/json.hpp>
#include <getopt.h>

using namespace std;
using json = nlohmann::json;

string generateId() {
    const char* hex = "0123456789abcdef";
    string id;
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(0, 15);
    for (int i=0; i<8; i++) id += hex[dis(gen)];
    return id;
}

string currentDate() {
    time_t t = time(nullptr);
    char buf[11];
    strftime(buf, sizeof(buf), "%Y-%m-%d", localtime(&t));
    return string(buf);
}

struct Habit {
    string id;
    string name;
    int frequencyDays;
    string lastDone;

    Habit() : frequencyDays(1) {}
    Habit(const string& n, int freq) : id(generateId()), name(n), frequencyDays(freq) {}

    string nextDueDate() const {
        if (lastDone.empty()) return currentDate();
        struct tm tm = {};
        strptime(lastDone.c_str(), "%Y-%m-%d", &tm);
        time_t t = mktime(&tm);
        t += frequencyDays * 24 * 3600;
        tm = *localtime(&t);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", &tm);
        return string(buf);
    }

    bool isDue() const {
        return nextDueDate() <= currentDate();
    }

    void markDone() {
        lastDone = currentDate();
    }
};

class Tracker {
private:
    vector<Habit> habits;
    string dataFile = "habits.json";

    void load() {
        ifstream f(dataFile);
        if (!f.is_open()) return;
        json j;
        f >> j;
        for (auto& item : j) {
            Habit h;
            h.id = item["id"];
            h.name = item["name"];
            h.frequencyDays = item["frequency_days"];
            h.lastDone = item["last_done"];
            habits.push_back(h);
        }
    }

    void save() {
        json j = json::array();
        for (auto& h : habits) {
            j.push_back({
                {"id", h.id},
                {"name", h.name},
                {"frequency_days", h.frequencyDays},
                {"last_done", h.lastDone}
            });
        }
        ofstream f(dataFile);
        f << setw(2) << j << endl;
    }

public:
    Tracker() { load(); }

    void add(const string& name, int freq) {
        Habit h(name, freq);
        habits.push_back(h);
        save();
        cout << "✅ Habit added: " << h.name << " (ID: " << h.id << ", frequency: " << freq << " days)\n";
    }

    void list() {
        if (habits.empty()) {
            cout << "No habits.\n";
            return;
        }
        cout << "\n📋 Your Habits:\n";
        for (auto& h : habits) {
            string due = h.nextDueDate();
            struct tm tmDue = {}, tmNow = {};
            strptime(due.c_str(), "%Y-%m-%d", &tmDue);
            strptime(currentDate().c_str(), "%Y-%m-%d", &tmNow);
            int daysUntil = (int)difftime(mktime(&tmDue), mktime(&tmNow)) / (24*3600);
            string status;
            if (daysUntil < 0) status = "🔔 DUE!";
            else if (daysUntil == 0) status = "⏰ due today";
            else if (daysUntil == 1) status = "due tomorrow";
            else status = "due in " + to_string(daysUntil) + " days";
            string last = h.lastDone.empty() ? "never" : h.lastDone;
            cout << "  " << h.id << ": " << h.name << " (every " << h.frequencyDays << " day" << (h.frequencyDays>1 ? "s" : "") << ") – last: " << last << ", next: " << due << " (" << status << ")\n";
        }
    }

    void check() {
        vector<Habit*> due;
        for (auto& h : habits) if (h.isDue()) due.push_back(&h);
        if (due.empty()) {
            cout << "✅ No habits due today. Well done!\n";
            return;
        }
        cout << "\n🔔 Due habits:\n";
        for (auto* h : due) {
            cout << "  " << h->id << ": " << h->name << " (frequency: " << h->frequencyDays << " days)\n";
            cout << "  Mark '" << h->name << "' as done? (y/n): ";
            string ans;
            getline(cin, ans);
            if (ans == "y" || ans == "Y") {
                h->markDone();
                cout << "  ✅ Habit '" << h->name << "' marked as done.\n";
            } else {
                cout << "  ⏳ Habit '" << h->name << "' skipped.\n";
            }
        }
        save();
    }

    void remove(const string& id) {
        for (auto it = habits.begin(); it != habits.end(); ++it) {
            if (it->id == id) {
                string name = it->name;
                habits.erase(it);
                save();
                cout << "✅ Habit '" << name << "' removed.\n";
                return;
            }
        }
        cout << "❌ Habit with ID " << id << " not found.\n";
    }

    void watch(int interval) {
        cout << "🔔 Habit Reminder – checking every " << interval << " minute(s). Press Ctrl+C to stop.\n";
        while (true) {
            vector<Habit*> due;
            for (auto& h : habits) if (h.isDue()) due.push_back(&h);
            if (!due.empty()) {
                time_t t = time(nullptr);
                tm* now = localtime(&t);
                char buf[6];
                strftime(buf, sizeof(buf), "%H:%M", now);
                cout << "\n🔔 Due habits at " << buf << ":\n";
                for (auto* h : due) cout << "  " << h->id << ": " << h->name << "\n";
            } else {
                cout << ".";
                cout.flush();
            }
            this_thread::sleep_for(chrono::minutes(interval));
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Usage: habit_reminder <command> [options]\n";
        return 1;
    }
    Tracker tracker;
    string cmd = argv[1];

    if (cmd == "add") {
        if (argc < 3) { cerr << "add <name> [--frequency N]\n"; return 1; }
        string name = argv[2];
        int freq = 1;
        for (int i=3; i<argc; i++) {
            if (string(argv[i]) == "--frequency" && i+1 < argc) {
                freq = stoi(argv[++i]);
            }
        }
        tracker.add(name, freq);
    } else if (cmd == "list") {
        tracker.list();
    } else if (cmd == "check") {
        tracker.check();
    } else if (cmd == "remove") {
        if (argc < 3) { cerr << "remove <id>\n"; return 1; }
        tracker.remove(argv[2]);
    } else if (cmd == "watch") {
        int interval = 5;
        for (int i=2; i<argc; i++) {
            if (string(argv[i]) == "--interval" && i+1 < argc) {
                interval = stoi(argv[++i]);
            }
        }
        tracker.watch(interval);
    } else {
        cerr << "Unknown command. Use add, list, check, remove, watch.\n";
        return 1;
    }
    return 0;
}
